src/beliefs.jl

```diff
@@ -135,7 +139,14 @@ function Random.rand(rng::AbstractRNG, b::WeightedParticleBelief)
     end
     return particles(b)[i]
 end
-Statistics.mean(b::WeightedParticleBelief) = dot(b.weights, b.particles)/weight_sum(b)
+# Statistics.mean(b::WeightedParticleBelief{T}) where {T <: Real} = dot(b.weights, b.particles) / weight_sum(b)
+# Statistics.mean(b::WeightedParticleBelief{T}) where {T <: Vector} = hcat(b.particles...) * b.weights / weight_sum(b)
+# Statistics.mean(b::WeightedParticleBelief{Vector{T}}) where {T <: Real} = sum(b.weights .* b.particles) / weight_sum(b)
+Statistics.mean(b::WeightedParticleBelief) = sum(b.weights .* b.particles) / weight_sum(b)
+function Statistics.cov(b::WeightedParticleBelief)
+    diff = hcat(b.particles...) .- mean(b)
```

Comment:
You should be able to stabilize the type by avoiding the splat operator and use `diff = reduce(hcat, b.particles) .- Statistics.mean(b)` instead. `reduce(::typeof(hcat), itr)` and `reduce(::typeof(vcat), itr)` have some special handling in `Base` to be fast and type stable. The same fix should help below to get rid of type instability.

Reply:
Also, you may not want to use diff as a variable name since it shadows `Base.diff` but that seems more or less unproblematic.

Reply:
Thanks for the comments! I will have a look at it in the next days!

Reply:
I have implemented a type stable method for everything now. There are now methods dispatching on beliefs where `T<:Number` and on beliefs where `T<:Vector`. This allows that `mean` always returns type `T` (otherwise it would return a one-element `Vector` also for `T<:Number`). In the same way `cov` returns `Number` instead of a one-element `Matrix`.